### PR TITLE
Convert devtools project JavaScript code to TypeScript

### PR DESCRIPTION
**Pull request by Gravatronics**

### CODE REVIEW

File: wcfsetup/install/files/js/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/AbstractPackageList.js [hunk 1]
```diff
@@ -1,272 +1,193 @@
 /**
```
Comment (TimWolla) [inline wcfsetup/install/files/js/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/AbstractPackageList.js:1]: Not reviewable.

File: wcfsetup/install/files/js/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/ExcludedPackages.js [hunk 2]
```diff
@@ -1,89 +1,56 @@
 /**
  * Manages the packages entered in a devtools project excluded package form field.
```
Comment (TimWolla) [inline wcfsetup/install/files/js/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/ExcludedPackages.js:2]: Not reviewable.

File: wcfsetup/install/files/js/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/OptionalPackages.js [hunk 3]
```diff
@@ -1,33 +1,26 @@
...
-                packageIdentifier: packageData.packageIdentifier
-            });
+    AbstractPackageList_1 = tslib_1.__importDefault(AbstractPackageList_1);
+    Core = tslib_1.__importStar(Core);
+    Language = tslib_1.__importStar(Language);
+    class OptionalPackages extends AbstractPackageList_1.default {
+        populateListItem(listItem, packageData) {
+            super.populateListItem(listItem, packageData);
+            listItem.innerHTML = ` ${Language.get("wcf.acp.devtools.project.optionalPackage.optionalPackage", {
+                packageIdentifier: packageData.packageIdentifier,
```
Comment (TimWolla) [inline wcfsetup/install/files/js/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/OptionalPackages.js:20]: You dropped `packageData.file`. Is this intentional?

File: wcfsetup/install/files/js/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/RequiredPackages.js [hunk 4]
```diff
@@ -1,102 +1,66 @@
 /**
  * Manages the packages entered in a devtools project required package form field.
```
Comment (TimWolla) [inline wcfsetup/install/files/js/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/RequiredPackages.js:2]: Not reviewable.

File: wcfsetup/install/files/js/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/Instructions.js [hunk 5]
```diff
@@ -1,626 +1,535 @@
 /**
  * Manages the instructions entered in a devtools project instructions form field.
```
Comment (TimWolla) [inline wcfsetup/install/files/js/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/Instructions.js:2]: Not reviewable.

File: wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/Instructions.ts [hunk 6]
```diff
@@ -0,0 +1,712 @@
...
+  protected readonly instructionsList: HTMLUListElement;
+  protected readonly instructionsType: HTMLSelectElement;
+  protected readonly instructionsTemplate: Template;
+  protected readonly instructionEditDialogTemplate: Template;
+  protected readonly pipDefaultFilenames: { [k: string]: string };
+
+  protected static applicationPips = ["acpTemplate", "file", "script", "template"];
+
+  // see `wcf\data\package\Package::isValidPackageName()`
+  protected static packageIdentifierRegExp = new RegExp(/^[a-zA-Z0-9_-]+\.[a-zA-Z0-9_-]+(\.[a-zA-Z0-9_-]+)+$/);
```
Comment (TimWolla) [inline wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/Instructions.ts:55]: Maybe `readonly` (I believe also applies to another file).

File: wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/Instructions.ts [hunk 7]
```diff
@@ -0,0 +1,712 @@
...
+  protected readonly formFieldId: string;
+  protected readonly fromVersion: HTMLInputElement;
+  protected instructionCounter = 0;
+  protected instructionsCounter = 0;
+  protected readonly instructionsEditDialogTemplate: Template;
+  protected readonly instructionsList: HTMLUListElement;
+  protected readonly instructionsType: HTMLSelectElement;
+  protected readonly instructionsTemplate: Template;
+  protected readonly instructionEditDialogTemplate: Template;
+  protected readonly pipDefaultFilenames: { [k: string]: string };
```
Comment (TimWolla) [inline wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/Instructions.ts:50]: Use a type alias / interface.

File: wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/Instructions.ts [hunk 8]
```diff
@@ -0,0 +1,712 @@
...
+
+  protected static applicationPips = ["acpTemplate", "file", "script", "template"];
+
+  // see `wcf\data\package\Package::isValidPackageName()`
+  protected static packageIdentifierRegExp = new RegExp(/^[a-zA-Z0-9_-]+\.[a-zA-Z0-9_-]+(\.[a-zA-Z0-9_-]+)+$/);
+
+  // see `wcf\data\package\Package::isValidVersion()`
+  protected static versionRegExp = new RegExp(
```
Comment (TimWolla) [inline wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/Instructions.ts:58]: readonly

File: wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/Instructions.ts [hunk 9]
```diff
@@ -0,0 +1,712 @@
...
+    /^([0-9]+).([0-9]+)\.([0-9]+)( (a|alpha|b|beta|d|dev|rc|pl) ([0-9]+))?$/i,
+  );
+
+  constructor(
+    formFieldId: string,
+    instructionsTemplate: Template,
+    instructionsEditDialogTemplate: Template,
+    instructionEditDialogTemplate: Template,
+    pipDefaultFilenames: { [k: string]: string },
```
Comment (TimWolla) [inline wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/Instructions.ts:67]: alias

File: wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/Instructions.ts [hunk 10]
```diff
@@ -0,0 +1,712 @@
...
+  /**
+   * Adds an instruction to a set of instructions as a consequence of the given event.
+   * If the instruction data is invalid, an error message is shown and no instruction is added.
+   */
+  protected addInstruction(event: Event): void {
+    event.preventDefault();
+    event.stopPropagation();
+
+    const instructionsId = ((event.currentTarget as HTMLElement).closest("li.section") as HTMLElement).dataset
+      .instrictionsId!;
```
Comment (TimWolla) [inline wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/Instructions.ts:129]: typo?

File: wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/Instructions.ts [hunk 11]
```diff
@@ -0,0 +1,712 @@
...
+    if (
+      !this.validateInstructionsType() ||
+      (this.instructionsType.value === "update" && !this.validateFromVersion(this.fromVersion))
+    ) {
+      return;
+    }
+
+    this.addInstructionsByData({
+      fromVersion: this.instructionsType.value === "update" ? this.fromVersion.value : "",
+      type: this.instructionsType.value as "install" | "update",
```
Comment (TimWolla) [inline wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/Instructions.ts:247]: Type alias for the union.

File: wcfsetup/install/files/js/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/OptionalPackages.js (hunk 3 above)
Comment (Gravatronics) [reply wcfsetup/install/files/js/WoltLabSuite/Core/Acp/Form/Builder/Field/Devtools/Project/OptionalPackages.js:20]: Yes.

**Review by TimWolla — APPROVED** [review]

**Pull request merged**